src/linear_algebra/quadratic_form.lean

```diff
@@ -381,11 +381,72 @@ end bilin_form
 namespace quadratic_form
 open bilin_form sym_bilin_form
 
+section associated'
+variables [invertible (2 : R)] {B₁ : bilin_form R M}
+
+/-- `associated'` is the additive homomorphism that sends a quadratic form to its associated
+symmetric bilinear form.  Over a commutative ring, use `associated`, which gives a linear map. -/
```

Comment:
Are you sure you need a second definition here? What if you change `associated` to have a second base ring for use in the linear map, with appropriate smul_comm_class arguments?

Reply:
Could you give a few more details? The issue is that there is no `has_scalar R (quadratic_form R M)` if `R` is not commutative, since `a * Q (b * x) ≠ b * b * (a * Q x)` if `a` and `b` do not commute.

Reply:
Sure, I've put the details in https://github.com/leanprover-community/mathlib/pull/6585#discussion_r589288048

Reply:
This sounds like it could be very elegant, but like @Vierkantor I'm not sure I quite follow the mechanics.  @eric-wieser would you mind rewriting the PR according to your idea after #6586 is merged?

Reply:
@eric-wieser I made an attempt at this myself, but got stuck at trying to allow the `S` to be implicit.  Do you have something in mind for this?  Does this require some `out_param` magic?

Reply:
Would you mind pushing your version to a branch?

Reply:
Sorry, I didn't keep the code!  And then I merged and it'll be an hour before there are new oleans.

Reply:
I don't think you can make `S` be implicit in any useful way, since all the lemmas are about the applied version anyway. I've pushed a change with an explicit `S`.

Reply:
Thanks for the revision!  This is somewhat more efficient, but I don't really like how the noncommutative version now requires a floating `ℕ` to get access to `associated'`.  Perhaps we need three versions, `associated_aux` (`S`-linear map), `associated'` (`→+`), and `associated` (`R`-linear map).  And we could get by with the definition only (no API), for the `S`-version, since its main purpose is to provide a simultaneous construction of the other two.  What do you think?

Reply:
I was starting to think maybe we should just have an unbundled `associated`, and have `associated_hom` as the `S`-bundled one. That way, we only need one lemma about `associated_hom` (that it applies to `associated`), and we don't end up with a duplicated version for R and S

Reply:
Note the `S` version is also possibly useful for obtaining a `rat`-linear map if `R` is a rat-algebra

Reply:
Alternatively, we can omit the use of the extra ring `S` altogether, but streamline the construction of `associated` by having it pass through `associated'`:
```lean
def associated : quadratic_form R₁ M →ₗ[R₁] bilin_form R₁ M :=
{ map_smul' := λ a Q, by { ext, simp, ring },
  .. quadratic_form.associated' }
```

Reply:
My claim is that the `S` version is useful by itself. Consider the (admitedly weird) case where `R₁` is the quaternions. With the `S` version, I can still recover the `real`-linear map, whereas with your suggestion I only have the additive version.

Reply:
> I was starting to think maybe we should just have an unbundled `associated`, and have `associated_hom` as the `S`-bundled one. That way, we only need one lemma about `associated_hom` (that it applies to `associated`), and we don't end up with a duplicated version for R and S

This would be ok for me (if we take "unbundled" to mean the `→+` version, since there is no point throwing out structure that is common to all three versions).  @Vierkantor, you wrote the file, what do you think?  Currently I think the fact that the operation respects scalar multiplication is used only once, in `quadratic_form.to_matrix'_smul`.

Reply:
Oh, by unbundled I actually mean the plain function, but I guess you're right, that approach should be fine too.

Reply:
Sorry to change my mind, but on reflection, I am not enthusiastic about losing structure from the most important case (`R` commutative, `S = R`).  I'd be interested to hear what other people think, too.

Alternative proposal: (1) `associated'` for `→+` with a full suite of lemmas; (2) `associated_hom` for `→ₗ[S]` with just the one lemma converting to `associated'`; (3) `associated` for `→ₗ[R]` with a full suite of lemmas.

Reply:
I have pushed a new approach, see [Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236585) for details.